vaadin-combo-box-flow/src/main/resources/META-INF/resources/frontend/comboBoxConnector.js

```diff
@@ -56,36 +79,59 @@ window.Vaadin.Flow.comboBoxConnector = {
       if (filterChanged) {
-        pageCallbacks = {};
         cache = {};
         lastFilter = params.filter;
+        this._debouncer = Debouncer.debounce(
+          this._debouncer,
+          timeOut.after(500),
+          () => {
+            clearPageCallbacks();
+            comboBox.clearCache();
+            if (params.filter === '') {
+              // Fixes the case when the filter changes
+              // from '' to something else and back to ''
+              // within debounce timeout, and the
+              // DataCommunicator thinks it doesn't need to send data
+              comboBox.$server.resetDataCommunicator();
+            }
+          });
+        return;
       }
 
       if (cache[params.page]) {
         // This may happen after skipping pages by scrolling fast
         commitPage(params.page, callback);
       } else {
-        const upperLimit = params.pageSize * (params.page + 1);
-
-        if (filterChanged) {
+        pageCallbacks[params.page] = callback
+        const activePages = Object.keys(pageCallbacks).map(page => parseInt(page));
+        const rangeMin = Math.min(...activePages);
+        const rangeMax = Math.max(...activePages);
+
+        if (activePages.length * params.pageSize > MAX_RANGE_COUNT) {
+          if (params.page === rangeMin) {
+            clearPageCallbacks([String(rangeMax)]);
+          } else {
+            clearPageCallbacks([String(rangeMin)]);
+          }
+          comboBox.dataProvider(params, callback);
+        } else if (rangeMax - rangeMin + 1 !== activePages.length) {
+          // Wasn't a sequential page index, clear the cache so combo-box will request for new pages
+          clearPageCallbacks();
+        } else {
+          // The requested page was sequential, extend the requested range
+          const startIndex = params.pageSize * rangeMin;
+          const endIndex = params.pageSize * (rangeMax + 1);
+          const count = endIndex - startIndex;
+          setRequestedRange = () => comboBox.$server.setRequestedRange(startIndex, count, params.filter);
+          if (!this._debouncer || !this._debouncer.isActive()) {
+            setRequestedRange();
+            setRequestedRange = null;
+          }
           this._debouncer = Debouncer.debounce(
             this._debouncer,
-            timeOut.after(500),
-            () => {
-              comboBox.$server.setRequestedRange(0, upperLimit, params.filter);
-              if (params.filter === '') {
-                // Fixes the case when the filter changes 
-                // from '' to something else and back to '' 
-                // within debounce timeout, and the
-                // DataCommunicator thinks it doesn't need to send data
-                comboBox.$server.resetDataCommunicator();
-              }
-            });
-        } else {
-          comboBox.$server.setRequestedRange(0, upperLimit, params.filter);
+            timeOut.after(50),
```

Comment:
IMO this could be increased to 100 or 200ms. It affects only fast scrolling and I think the benefit of loading less data overcomes the downside of added lag.

Reply:
Done.